### DISCUSSION TITLE
Make it easier to use subsets in dynamic enums

### DISCUSSION THREAD

**Discussion by cmungall** [linkml/linkml#2300] [Ideas]
From slack

The dynamic enums (https://linkml.io/linkml/schemas/enums.html#dynamic-enums) are a flexible way to specify boolean combination of branches of an ontology. They can be materialized using OAK but there is no dependence on any tool, the spec is designed to be simple enough that anyone can develop a value set expander or validator using their API or ontology library of choice.

Currently the spec language is intentionally limited to make it easier to have multiple implementations. For example, you can't specify complex lexical patterns over custom combinations of predicates (even though this might be possible directly with OAK). Generally it's just boolean combos of simple graph queries.

This means there is no way to restrict a value set to an ontology subset. We could add support for this in one of two ways:

 - allow adding a subset predicate (e.g. oboinowl:inSubset) to the list of traversal predicates
 - extending the metamodel to allow for an `in_subset` metaslot. It would be up to individual tools to allow the user a mechanism to bind this to a specific property (since different ontologies may model in_subset differently)

I think the 2nd is preferable. This does add an extra burden on implementors of value set expansion and validation tools. But I think the subset use case is important enough to warrant this.

Some examples

- GO "slims" use oboInOwl:inSubset
- CL, Uberon, Mondo, many other ontologies also uses this
- NCIT uses its own property but this is normalized to inSubset in the OboInOwl version

See also https://github.com/information-artifact-ontology/ontology-metadata/issues/80

**Comment by cmungall** [comment]
From @shahimessaid:

https://confluence.ihtsdotools.org/display/DOCEXTPG/4.2+Modules

**Comment by ShahimEssaid** [comment]
A few pointers based on our discussion:

R4:

ValueSet:
https://hl7.org/fhir/R4/valueset.html#compositions (and the rest of the page)
https://hl7.org/fhir/R4/valueset-definitions.html#ValueSet.compose.include.filter

CodeSystem properties (can be used for value set "tagging" based on the native source vocabulary):
https://hl7.org/fhir/R4/codesystem.html#properties

R5 has better documentation:
https://hl7.org/fhir/valueset.html#compositions

Development version might be even better (upcoming R6):
https://build.fhir.org/valueset.html#compositions

Other possibly interesting pages:
https://build.fhir.org/terminologies.html
https://terminology.hl7.org/SNOMEDCT.html#snomed-ct-properties
